src/unsafe-keyword.md

```diff
@@ -0,0 +1,79 @@
+# The `unsafe` keyword
+
+The `unsafe` keyword in Rust can occur in several different context: unsafe
+functions (`unsafe fn`), unsafe blocks (`unsafe {}`), unsafe traits (`unsafe
+trait`), and unsafe trait implementations (`unsafe impl`). It plays several
+different roles, depending on where it is used and whether the
+`unsafe_op_in_unsafe_fn` lint is enabled:
+- it is used to mark code that *defines* extra safety conditions (`unsafe fn`,
+  `unsafe trait`)
+- it is used to mark code that needs to *satisfy* extra safety conditions
+  (`unsafe {}`, `unsafe impl`, `unsafe fn` without `unsafe_op_in_unsafe_fn`)
+
+The following discusses each of these cases. See the
+[keyword documentation][keyword] for some illustrative examples.
+
+[keyword]: ../std/keyword.unsafe.html
+
+## Unsafe functions (`unsafe fn`)
+
+Unsafe functions are functions that are not safe in all contexts and/or for all
+possible inputs. We say they have *extra safety conditions*, which are
+requirements that must be upheld by all callers and that the compiler does not
+check. For example, `get_unchecked` has the extra safety
+condition that the index must be in-bounds. The module defining an unsafe
+function is responsible for documenting what those extra safety conditions are.
```

Comment:
This seems rather specific to say it must go in the *module* documentation. For example, the aforementioned `get_unchecked` is just documented on the method itself, not in the module docs. Can this be reworded to be a little more general?

Reply:
I'm going to say that we should point out that this isn't an actual requirement of the language, but rather expected practice. There's nothing in `rustc` that will (or could) stop your code from compiling because you failed to explain what extra conditions you are imposing.

Reply:
I didn't mean this to imply it must be the module-level docs, just 'somewhere inside the module'.